### DISCUSSION TITLE
Supporting a sequence of binarizers?

### DISCUSSION THREAD

**Discussion by duanqn** [zxing-cpp/zxing-cpp#501] [Ideas]
Currently there are 4 binarizers in ZXing-cpp, and they each have strengths & weaknesses on certain images. I think a way to increase recognition success rate is to chain multiple binarizers together, and it is best done inside ZXing-cpp. I made a draft PR to demonstrate the concept. (#500)
Unfortunately this change does break the ABI interface.
Moving forward, it is also possible for us to take more binarization methods. For example, I noticed this repo https://github.com/opencv/opencv_contrib/tree/4.x/modules/wechat_qrcode has some additional binarization methods.

**Comment by axxel** [comment]
Thanks for starting this discussion. I had a look at your PR but don't think that this would be the best way forward. Based on the feedback via issue reports and looking at how other projects use this library, I have the feeling that adding a binarizer set option would effectively not be really beneficial to the user base as a whole. The average user would have absolutely no way of knowing which is best for its situation.

That said, I can definitively see the possibility/need for improving the detection rate as far as the binarization is concerned. What I believe we should try to do is simply improve the (default) binarizer with that regard.

What I can also imagine as being relatively straight forward is to add a `Binarizer::Automatic` option that would then cause the library to try to find the best possible binarization, potentially trying multiple in a row. Right now trying multiple options would only be meaningful for matrix codes and the only two meaningful ones to try would be `LocalAverage` and `GlobalHistrogram`. The effect on the runtime would also be relatively limited in a "search for all symbologies" situation right now, as the most time is spent in the linear barcodes search.

If you have a superior binarizer alternative (like any of the wechat options), please let me know. Maybe we can then simply replace one or both of the two mentioned ones. Superior could mean "detects more images" or being faster.

**Reply by duanqn** [reply]
> The average user would have absolutely no way of knowing which is best for its situation.

Agreed. I don't know what's the best binarizer either, so I thought I can use all of them together for enhanced robustness.

> What I believe we should try to do is simply improve the (default) binarizer with that regard.

That sounds like a nice idea, but
1) I'm not sure if it is possible to build a good "universal binarizers" for all scenarios. My opinion is that different binarization methods will each have their suitable scenario. And if we want to build a generic solution, we still have to iterate through different binarization methods.
2) It is going to be a fair amount of work, tuning the heuristic with different samples. As this is an open source project, we can't count on people to do this work & maintain it in the long run.

> Right now trying multiple options would only be meaningful for matrix codes and the only two meaningful ones to try would be LocalAverage and GlobalHistrogram.

In fact I'm also using the ThresholdBinarizer with these two. I noticed that, a few reported issues (#437, #479) can be solved by "hacking" a threshold into the ThresholdBinarizer, so I thought it would be beneficial to try different thresholds in the binarization phase. Of course, the current ThresholdBinarizer has a fixed threshold of 128 so I cannot do this right now.

My rationale for doing this inside ZXing-cpp is to share some progress between different binarizers (e.g. converting RGB image to gray scale / avoid running the 1D part of both LocalAverage and GlobalHistrogram binarizer).

About the wechat binarizers -- They were implemented on the other zxing-cpp project. I considered porting them here, but their binarizers contain a few calls to OpenCV functions. What's your opinion on introducing OpenCV as a dependency? If you are OK with using OpenCV then it would be easier to port them. Note that wechat is heavily focusing on QRCode.

**Reply by axxel** [reply]
>     1. I'm not sure if it is possible to build a good "universal binarizers" for all scenarios. My opinion is that different binarization methods will each have their suitable scenario. And if we want to build a generic solution, we still have to iterate through different binarization methods.

Well, I guess that depends on the performance of the best binarizer available. Right now we can obviously get a select few more samples to work when using both `LocalAverage` and `GlobalHistogram`. A new binarizer (maybe based on one of the wechat opencv contributions) may be better than both.

>     2. It is going to be a fair amount of work, tuning the heuristic with different samples. As this is an open source project, we can't count on people to do this work & maintain it in the long run.

Well... I'm doing that for the whole library for about 4 years now. :)


> In fact I'm also using the ThresholdBinarizer with these two. I noticed that, a few reported issues (#437, #479) can be solved by "hacking" a threshold into the ThresholdBinarizer, so I thought it would be beneficial to try different thresholds in the binarization phase. Of course, the current ThresholdBinarizer has a fixed threshold of 128 so I cannot do this right now.

Well, you could of course try the `ThresholdBinarizer` with all 255 possible values but that would certainly not have an acceptable cost benefit ratio in my opinion.

> My rationale for doing this inside ZXing-cpp is to share some progress between different binarizers (e.g. converting RGB image to gray scale / avoid running the 1D part of both LocalAverage and GlobalHistrogram binarizer).

I understand.

> About the wechat binarizers -- They were implemented on the other zxing-cpp project. I considered porting them here, but their binarizers contain a few calls to OpenCV functions. What's your opinion on introducing OpenCV as a dependency? If you are OK with using OpenCV then it would be easier to port them. Note that wechat is heavily focusing on QRCode.

I would definitely not add a build/runtime dependency on OpenCV. But if you tried one of those and it should turn out to be a substantial improvement, then it should be possible and worthwhile to port / re-implement the underlying algorithm. I do for example have an Otsu-based binarizer code here that did unfortunately not improve the situation sufficiently in my view.
